engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EqualsAnyNode.java

```diff
@@ -0,0 +1,609 @@
+      throw new RuntimeException(e);
+    }
+  }
+
+  /** Equals for Atoms and AtomConstructors */
+
+  @Specialization
+  boolean equalsAtomConstructors(AtomConstructor selfConstructor, AtomConstructor otherConstructor) {
+    return selfConstructor == otherConstructor;
+  }
+
+  /**
+   * How many {@link EqualsAnyNode} should be created for fields in specialization for atoms.
+   */
+  static final int equalsNodeCountForFields = 10;
+
+  static EqualsAnyNode[] createEqualsNodes(int size) {
+    EqualsAnyNode[] nodes = new EqualsAnyNode[size];
+    Arrays.fill(nodes, EqualsAnyNode.build());
+    return nodes;
+  }
+
+  static InvokeEqualsNode[] createInvokeEqualsNodes(int size) {
+    InvokeEqualsNode[] nodes = new InvokeEqualsNode[size];
+    Arrays.fill(nodes, InvokeEqualsNode.build());
+    return nodes;
+  }
+
+  @Specialization
+  boolean equalsAtoms(Atom self, Atom other,
+      @Cached LoopConditionProfile loopProfile,
+      @Cached(value = "createEqualsNodes(equalsNodeCountForFields)", allowUncached = true) EqualsAnyNode[] fieldEqualsNodes,
+      @Cached(value = "createInvokeEqualsNodes(equalsNodeCountForFields)", allowUncached = true) InvokeEqualsNode[] fieldInvokeEqualsNodes,
+      @Cached InvokeEqualsNode atomInvokeEqualsNode,
+      @Cached ConditionProfile enoughEqualNodesForFieldsProfile,
+      @Cached ConditionProfile fieldsSizeNotEqualProfile,
+      @Cached ConditionProfile constructorsNotEqualProfile) {
+    if (fieldsSizeNotEqualProfile.profile(
+        self.getFields().length != other.getFields().length
+    )) {
+      return false;
+    }
+    if (constructorsNotEqualProfile.profile(
+        self.getConstructor() != other.getConstructor()
+    )) {
+      return false;
+    }
+
+    if (self instanceof Atom selfAtom
+        && other instanceof Atom otherAtom
+        && atomOverridesEquals(selfAtom)) {
+      return atomInvokeEqualsNode.execute(selfAtom, otherAtom);
+    }
+
+    assert self.getFields().length == other.getFields().length;
+    int fieldsSize = self.getFields().length;
+    if (enoughEqualNodesForFieldsProfile.profile(fieldsSize <= equalsNodeCountForFields)) {
+      loopProfile.profileCounted(fieldsSize);
+      for (int i = 0; loopProfile.inject(i < fieldsSize); i++) {
+        boolean areFieldsSame;
+        // Note that otherFieldAtom does not have to override `==` method, it is sufficient if
+        // selfFieldAtom overrides `==` method.
+        Object selfField = self.getFields()[i];
+        Object otherField = other.getFields()[i];
+        if (selfField instanceof Atom selfFieldAtom
+            && otherField instanceof Atom otherFieldAtom
+            && atomOverridesEquals(selfFieldAtom)
+        ) {
+          areFieldsSame =
+              fieldInvokeEqualsNodes[i].execute(selfFieldAtom, otherFieldAtom);
+        } else {
+          areFieldsSame = fieldEqualsNodes[i].execute(
+              selfField,
+              otherField
+          );
+        }
```

Comment:
Why are we doing this shortcut for invoking `==` on nested Atoms instead of just calling the `fieldEqualsNode` and allowing it to choose which approach to delegate to?

Reply:
It could be possible to get rid of this. This just seems like a leftover now.